### DISCUSSION TITLE
[QUESTION] How to automatically validate input parameter as IPv4Address

### DISCUSSION THREAD

**Discussion by johnthagen** [fastapi/typer#1310] [Questions] [answered]
### First check

* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the Typer documentation, with the integrated search.
* [x] I already searched in Google "How to X in Typer" and didn't find any information.
* [x] I already searched in Google "How to X in Click" and didn't find any information.

### Description

With `argparse` it it possible to do something like:

```py
    parser.add_argument(
        "--host",
        type=IPv4Address,
        default=IPv4Address("127.0.0.1"),
    )
```

This will automatically try to construct a `IPv4Address` type from the input and validate that it is a valid `IPv4Address`.

```
$ main.py --host=a
...
main.py: error: argument --host: invalid IPv4Address value: 'a'
```

I didn't see this type documented in `typer`. Is this possible to do?

**Comment by johnthagen** [comment]
Perhaps this should be a feature request.

```py
@app.command()
def main(
    host: IPv4Address = typer.Option(
        IPv4Address("127.0.0.1"), help="Host interface to listen on."
    ),
) -> None:
    pass
```

Output when run:

```
RuntimeError: Type not yet supported: <class 'ipaddress.IPv4Address'>
```

**Comment by daddycocoaman** [comment]
In the meantime, you can get use the `callback` parameter to call `IPv4Address`. If it's valid, cool. If not, it'll error.

```python
@app.command()
def main(host: stre= typer.Option("127.0.0.1", help="Host interface to listen on.", callback=IPv4Address)) -> None:
    pass
```

**Comment by johnthagen** [comment]
@daddycocoaman Thanks, this is close to what I was looking for. The `callback=IPv4Address` does what I expect, but then the type hint is incorrect. It is type hinted as `str` but what is actually passed in the `host` variable is of type `IPv4Address`. Ideally, for type checking and better editor integration, we'd like them to match.

**Comment by baterflyrity** [comment]
+1, I need this feature in 99% of cli projects.

**Comment by baterflyrity** [comment]
@johnthagen , unfortunately, click (basic library) seem to not support IPv4Address type. So guess we need to request this type on click's issue tracker.

p.s.
As you can [see](https://github.com/tiangolo/typer/blob/c3a4c72a4f1a18b7f16addb508862ec63411fa63/typer/main.py#L503) types are just mapped to underlying library ones.

**Comment by daddycocoaman** [comment]
Looks like creating new Click types isn't that difficult either, although
it seems that the Click team isn't always open integrating new types.  This
guy tried to get Date merged into Click and it was rejected.

https://github.com/pallets/click/issues/1655

It might be worth creating an additional package for extra click types that
Typer can import,
or building on the existing project for additional parameters.

https://github.com/click-contrib/click_params

**Comment by YuriiMotov** [comment, marked as answer]
You can make `IPv4Address` (and other custom types) working by specifying `parser` parameter:

```py
from ipaddress import IPv4Address

import typer
from typing_extensions import Annotated

def main(
    host: Annotated[
        IPv4Address,
        typer.Argument(
            parser=IPv4Address,
            default_factory=lambda: IPv4Address("127.0.0.1"),
            help="Host address (IP v4)",
            show_default="127.0.0.1",
        ),
    ],
):
    print(f"host is {host}")

if __name__ == "__main__":
    typer.run(main)
```

Docs: https://typer.tiangolo.com/tutorial/parameter-types/custom-types/
